data/blog/enjoyable-database-testing.md

```diff
@@ -0,0 +1,277 @@
+    delete conn theId
+    list conn `shouldReturn` before
+```
+
+There are still potential problems with this modification.
+
+At the end of the day it might not be clear how to make the test robust.
+
+One way we can regain the isolation of separate database clusters and `postgres` instances is by using the databases mechanisms for query isolation.
+
+To faciliate query isolation we'll write an function to wrap a list of statements in a transaction that we rollback instead of committing:
+
+```haskell
+abort :: (Connection -> IO a) -> Connection -> IO a
+abort f conn = bracket_
+  (execute_ conn "BEGIN")
+  (execute_ conn "ROLLBACK")
+  (f conn)
+```
+
+We can now prefix our tests with abort and they will not interfer with each other
+
+```haskell
+aroundAll withSetup $ describe "list/add/delete" $ do
+  it "return [] initially" $ withPool $ abort $ \conn -> do
+    list conn `shouldReturn` []
+  it "returns the created elements" $ withPool $ abort $ \conn -> do
+    theId <- add conn 1
+    list conn `shouldReturn` [theId]
+  it "deletes what is created" $ withPool $ abort $ \conn -> do
+    theId <- add conn 2
+    delete conn theId
+    list conn `shouldReturn` []
+```
+
+It is worth pointing out that not every PostgreSQL sql statement can run in a transaction. Additionally there is no isolation level that can bring the database to the same state the way starting at a cluster snapshot can. Things like series are incremented and not decrememented on rollback among other MVCC infidelities. That said the fast majority of queries are perfectly isolated using  transaction and `abort` will be sufficent to return the database to a original state (for all intensive purposes).
+
+## The Pleasure and Pain of `parallel`
+
+Database queries are meant to be run in parallel with other queries. We can take advantage of this by running our tests in parallel to improve performance.
+
+We can change our test to run in parallel by adding the `parallel` combinator:
+
+```haskell
+aroundAll withSetup $ describe "list/add/delete" $ parallel $ do
+  it "return [] initially" $ withPool $ abort $ \conn -> do
+    list conn `shouldReturn` []
+  it "returns the created elements" $ withPool $ abort $ \conn -> do
+    theId <- add conn 1
+    list conn `shouldReturn` [theId]
+  it "deletes what is created" $ withPool $ abort $ \conn -> do
+    theId <- add conn 2
+    delete conn theId
+    list conn `shouldReturn` []
+```
+
+Unfortunately if we run our tests again we will notice failures.
+
+The problem is our `abort` function is wrapping everything in a `READ_COMMITTED` isolation level. This is equivalent to each statement receiving it's own snapshot of the database but we need the entire transaction to have a single consistent snapshot of the database so we need to use either `REPEATABLE_READ` or `SERIALIZABLE` isolation.
+
+In our particular example the solution is to use the higher isolation and retry on serializable errors. However this is not always possible.
```

Comment:
> In our particular example the solution is to use the higher isolation and retry on serializable errors. However this is not always possible.

In our particular example using modifying `abort` to use `SERIALIZABLE` would allow solve our parallel test issues.